### ISSUE TITLE
How to checkout a replaced commit hash?

### ISSUE THREAD

**Issue by gautelo** [newren/git-filter-repo#229]
After running `git filter-repo --path Subfolder --path .gitattributes --path .gitignore` on a repo clone, I'm left with a nice clean repo for that sub folder. Exactly what I want - great! But I notice that the commit IDs have been rewritten, and I have some references to these that require me to keep track of that change. After some investigation I see that that's a necessity, but that git-replace is supposed to provide the solution. Sadly I don't understand how it's supposed to work or if I'm simply missing a step.

If I run `git replace --list --format=long` I get something like the following...
```
01ba5a56f0fed0328b4630e68dd3974a6eb5e29b ((null)) -> 1291311faacf5e77757f15e22049f30ea14d45ce (commit)
0c24ecfb9c2adc5b2afc69f068b94d56c5bf2178 ((null)) -> 73f023da03db249886047513782fa3a72c7124dc (commit)
17b193cbcee1264daf96c9b6f8fd6bf361562048 ((null)) -> d87a8dc11970343a3dca6f3899703e41c26d56bc (commit)
```
If I now ask if `git branch --contains 01ba5a56f0fed0328b4630e68dd3974a6eb5e29b` it says no, though it succeeds for the replacing commit 1291311faacf5e77757f15e22049f30ea14d45ce. I'm wondering if perhaps the reason is that it doesn't know that that SHA1 is a commit object, since it says ((null)) instead of (commit).

I also thought I'd find these mapped under .git/refs/replace/, but it's empty. I suppose they may have been packed since they are mentioned in both .git/packaged-refs and .git/info/refs. But frankly, I don't know git well enough to know what's what in here. I've tried to keep out of the .git folder as much as I can. Seems those days are over though.

Furthermore, when I push my repo and make a fresh clone there are no `git replace --list` entries at all. This leads me to believe I should push something somehow.

So, my question is this:
_What must I do in order to be able to do a `git checkout <replaced-sha1>` and other git operations on such replaced-sha1's?_

**Comment by newren** [comment]
I'm surprised that `git branch --contains <replaced-hash>` doesn't give an answer.  I can duplicate, though.  Seems like a bug; I'll report it to the other git developers.  Thanks for reporting it.

`git checkout <replaced-hash>`, however, will check out the relevant new hash.  `git log <replaced-hash>` will show the replaced commit's history.  `git diff <replaced-hash1> <replaced-hash2>` will show the diff of the new versions of the two replaced commits.  So most git commands accept replaced-hashes just fine.  (I'm trying to train myself to use "hash" instead of "sha1" due to the ongoing sha256 migration.)

git clone will only clone references under `refs/heads/*` and `refs/tags/*` by default.  Things like notes, replace refs, PR metadata, etc., are not included.  See bullet point 6 in the "DISCUSSION" section of the manual (https://htmlpreview.github.io/?https://github.com/newren/git-filter-repo/blob/docs/html/git-filter-repo.html#DISCUSSION) for instructions on making people be able to clone them.

**Comment by newren** [comment]
Upstream report of the `git branch` issue: https://lore.kernel.org/git/CABPp-BEAbN05+hCtK=xhGg5uZFqbUvH9hMcCNMcBWp5JWLqzPw@mail.gmail.com/T/#u

**Comment by gautelo** [comment]
Thank you for looking into this @newren. `git checkout <replaced-hash>` does indeed execute, but like you mentioned in your upstream report, it doesn't work correctly either, unfortunately.

I don't have any replace refs in my repo from beforhand, so I'm not concerned with that. But the replace refs created by filter-repo do not seem to be pushed upstream. Let me outline what I attempted, and what I had to do to solve it for anyone else who might have the same issue. Basically, what I'm trying is this:

```
git clone https://myuser@bitbucket.org/myworkspace/source-repo.git SubFolder
cd SubFolder
git filter-repo --path SubFolder --path .gitattributes --path .gitignore
git remote add origin https://myuser@bitbucket.org/myworkspace/subfolder.git
git push -u origin --all
cd ..
git clone https://myuser@bitbucket.org/myworkspace/subfolder.git SubFolderClone
```

When I run `git replace --list --format=long` in SubFolder I get the expected entries as shown in the original post, but when I do it in SubFolderClone I don't get any. So basically I'm losing the replace refs filter-repo has created for me.

How can I keep them? Your last paragraph clued me in to what I was doing wrong. 

Not only was I not pulling the replace refs from the remote when fetching, which is solved by either adding git config for always fetching them, as suggested in the discussion you linked or by explicitly fetching them. It appears that replace refs aren't pushed by default either. Hence doing a `git push origin 'refs/replace/*'` was necessary for my remote to receive the replacements.

I might have missed it, but I didn't notice anything in the docs mentioning having to explicitly push the replace refs in the documentation. May I suggest adding it @newren?

Turns out I was right initially 

> This leads me to believe I should push something somehow.

In summary doing the following ensures you keep the replace refs
```
git push origin 'refs/replace/*'
```
and either
```
git fetch origin 'refs/replace/*:refs/replace/*'
```
or 
```
git config --add remote.origin.fetch +refs/replace/*:refs/replace/*
git fetch
```

This means that once a fix is provided for the issue filed by @newren above, things should work as expected.
